### PR TITLE
Fix QR code login support in rust

### PR DESCRIPTION
**Pull request by BillCarsonFr**
## Type of change

- [ ] Feature
- [x] Bugfix
- [ ] Technical
- [ ] Other :

## Content

Fixes https://github.com/vector-im/element-android/issues/8653
Depends on new rust sdk  https://github.com/matrix-org/matrix-rust-sdk/pull/2632

## Tests

- Step 1
- Step 2
- Step ...

## Tested devices

- [ ] Physical
- [ ] Emulator
- OS version(s):

## Checklist

- [ ] Changes has been tested on an Android device or Android emulator with API 21
- [ ] UI change has been tested on both light and dark themes
- [ ] Accessibility has been taken into account. See https://github.com/vector-im/element-android/blob/develop/CONTRIBUTING.md#accessibility
- [ ] Pull request is based on the develop branch
- [ ] Pull request includes a new file under ./changelog.d. See https://github.com/vector-im/element-android/blob/develop/CONTRIBUTING.md#changelog
- [ ] Pull request includes screenshots or videos if containing UI changes
- [ ] Pull request includes a [sign off](https://matrix-org.github.io/synapse/latest/development/contributing_guide.html#sign-off)
- [ ] You've made a self review of your PR
- [ ] If you have modified the screen flow, or added new screens to the application, you have updated the test [UiAllScreensSanityTest.allScreensTest()](https://github.com/vector-im/element-android/blob/main/vector/src/androidTest/java/im/vector/app/ui/UiAllScreensSanityTest.kt#L73)

### CODE REVIEW

File: matrix-sdk-android/src/kotlinCrypto/java/org/matrix/android/sdk/internal/crypto/SecretShareManager.kt [hunk 1]
```diff
@@ -136,6 +136,13 @@ internal class SecretShareManager @Inject constructor(
                             .w("handleSecretRequest() : malformed request norequestingDeviceId ")
                 }
 
+        if (request.requestingDeviceId == credentials.deviceId) {
+            return Unit.also {
+                Timber.tag(loggerTag.value)
+                        .v("handleSecretRequest() : Ignore request from self device")
```
Comment (bmarty) [inline matrix-sdk-android/src/kotlinCrypto/java/org/matrix/android/sdk/internal/crypto/SecretShareManager.kt:142]: Is it possible to receive a toDevice for ourself?

File: matrix-sdk-android/src/kotlinCrypto/java/org/matrix/android/sdk/internal/crypto/SecretShareManager.kt [hunk 2]
```diff
@@ -136,6 +136,13 @@ internal class SecretShareManager @Inject constructor(
                             .w("handleSecretRequest() : malformed request norequestingDeviceId ")
                 }
 
+        if (request.requestingDeviceId == credentials.deviceId) {
```
Comment (bmarty) [inline matrix-sdk-android/src/kotlinCrypto/java/org/matrix/android/sdk/internal/crypto/SecretShareManager.kt:139]: Maybe
```suggestion
        if (deviceId == credentials.deviceId) {
```

File: matrix-sdk-android/src/kotlinCrypto/java/org/matrix/android/sdk/internal/crypto/SecretShareManager.kt [hunk 3]
```diff
@@ -254,6 +261,37 @@ internal class SecretShareManager @Inject constructor(
...
+
+        secretNames.forEach { secretName ->
+            val toDeviceContent = SecretShareRequest(
+                    requestingDeviceId = credentials.deviceId,
+                    secretName = secretName,
+                    requestId = createUniqueTxnId()
+            )
+
+            val contentMap = MXUsersDevicesMap<Any>()
+            contentMap.setObject(credentials.userId, "*", toDeviceContent)
```
Comment (bmarty) [inline matrix-sdk-android/src/kotlinCrypto/java/org/matrix/android/sdk/internal/crypto/SecretShareManager.kt:276]: Ah, OK, we send the toDevice to `*`.

File: matrix-sdk-android/src/main/java/org/matrix/android/sdk/api/rendezvous/Rendezvous.kt [hunk 4]
```diff
@@ -225,12 +221,7 @@ class Rendezvous(
             // request secrets from the verifying device
             Timber.tag(TAG).i("Requesting secrets from $verifyingDeviceId")
 
-            session.sharedSecretStorageService().let {
-                it.requestSecret(MASTER_KEY_SSSS_NAME, verifyingDeviceId)
-                it.requestSecret(SELF_SIGNING_KEY_SSSS_NAME, verifyingDeviceId)
-                it.requestSecret(USER_SIGNING_KEY_SSSS_NAME, verifyingDeviceId)
-                it.requestSecret(KEYBACKUP_SECRET_SSSS_NAME, verifyingDeviceId)
-            }
+            session.sharedSecretStorageService().requestMissingSecrets()
```
Comment (bmarty) [inline matrix-sdk-android/src/main/java/org/matrix/android/sdk/api/rendezvous/Rendezvous.kt:224]: I understand that we request the secrets from all our devices, and not the device which was used to verify (`verifyingDeviceId`). I think it's fine, but maybe update the comment and the log above?

File: matrix-sdk-android/src/rustCrypto/java/org/matrix/android/sdk/internal/crypto/SecretShareManager.kt [hunk 5]
```diff
@@ -16,15 +16,33 @@
...
+        }
+    }
+
+    suspend fun requestMissingSecrets() {
+        this.olmMachine.get().requestMissingSecretsFromOtherSessions()
+
+        // immediately send the requests
+        outgoingRequestsProcessor.processOutgoingRequests(this.olmMachine.get()) {
+            it is Request.ToDevice && it.eventType == EventType.REQUEST_SECRET
+        }
```
Comment (bmarty) [inline matrix-sdk-android/src/rustCrypto/java/org/matrix/android/sdk/internal/crypto/SecretShareManager.kt:46]: I do not really understand why the two method impl. are different. There is a if at line 32. Not applicable in this one?

File: matrix-sdk-android/src/main/java/org/matrix/android/sdk/internal/crypto/secrets/DefaultSharedSecretStorageService.kt [hunk 6]
```diff
@@ -385,7 +385,12 @@ internal class DefaultSharedSecretStorageService @Inject constructor(
         return IntegrityResult.Success(keyInfo.content.passphrase != null)
     }
 
+    @Deprecated("Requesting custom secrets not yet support by rust stack, prefer requestMissingSecrets")
```
Comment (bmarty) [inline matrix-sdk-android/src/main/java/org/matrix/android/sdk/internal/crypto/secrets/DefaultSharedSecretStorageService.kt:388]: I do not think it is useful to have the `@Deprecated` here. It's in the interface.

File: vector/src/test/java/im/vector/app/test/fakes/FakeSharedSecretStorageService.kt [hunk 7]
```diff
@@ -75,10 +75,15 @@ class FakeSharedSecretStorageService : SharedSecretStorageService by mockk() {
 
     override fun checkShouldBeAbleToAccessSecrets(secretNames: List<String>, keyId: String?) = integrityResult
 
+    @Deprecated("Requesting custom secrets not yet support by rust stack, prefer requestMissingSecrets")
```
Comment (bmarty) [inline vector/src/test/java/im/vector/app/test/fakes/FakeSharedSecretStorageService.kt:78]: I do not think it is useful to have the `@Deprecated` here. It's in the interface, and for Fake we do not really care.

**Review by bmarty — APPROVED** [review]
LGTM, some non-blocking remarks/questions.

File: matrix-sdk-android/src/main/java/org/matrix/android/sdk/internal/crypto/secrets/DefaultSharedSecretStorageService.kt (hunk 6 above)
Comment (BillCarsonFr) [reply matrix-sdk-android/src/main/java/org/matrix/android/sdk/internal/crypto/secrets/DefaultSharedSecretStorageService.kt:388]: ktlint complains if it's not there also

File: vector/src/test/java/im/vector/app/test/fakes/FakeSharedSecretStorageService.kt (hunk 7 above)
Comment (BillCarsonFr) [reply vector/src/test/java/im/vector/app/test/fakes/FakeSharedSecretStorageService.kt:78]: see above

File: matrix-sdk-android/src/main/java/org/matrix/android/sdk/api/rendezvous/Rendezvous.kt (hunk 4 above)
Comment (BillCarsonFr) [reply matrix-sdk-android/src/main/java/org/matrix/android/sdk/api/rendezvous/Rendezvous.kt:224]: updated

File: matrix-sdk-android/src/rustCrypto/java/org/matrix/android/sdk/internal/crypto/SecretShareManager.kt (hunk 5 above)
Comment (BillCarsonFr) [reply matrix-sdk-android/src/rustCrypto/java/org/matrix/android/sdk/internal/crypto/SecretShareManager.kt:42]: right, will call the other one and add warning that rust stak doesn't support requesting custom secrets yet

**Review by bmarty — APPROVED** [review]
Thanks for the update!

**Pull request merged**